File: app/Models/Product.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Builder;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Database\Eloquent\SoftDeletes;

class Product extends Model
{
    use SoftDeletes;

    protected $fillable = [
        'user_id',
        'name',
        'stock',
        'cost_price',
        'selling_price',
        'image_path',
    ];

    protected function casts(): array
    {
        return [
            'stock' => 'integer',
            'cost_price' => 'decimal:2',
            'selling_price' => 'decimal:2',
        ];
    }

    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }

    public function transactionItems(): HasMany
    {
        return $this->hasMany(TransactionItem::class);
    }

    public function scopeOwnedBy(Builder $query, User|int $user): Builder
    {
        $userId = $user instanceof User ? $user->id : $user;

        return $query->where('user_id', $userId);
    }
}

File: mobile_app/lib/domain/entities/product.dart
class Product {
  const Product({
    required this.id,
    required this.name,
    required this.stock,
    required this.minimumStock,
    required this.costPrice,
    required this.sellingPrice,
    required this.stockStatus,
    required this.stockStatusLabel,
    required this.needsRestock,
    this.imagePath,
    this.imageUrl,
    this.createdAt,
    this.updatedAt,
  });

  final int id;
  final String name;
  final int stock;
  final int minimumStock;
  final double costPrice;
  final double sellingPrice;
  final String stockStatus;
  final String stockStatusLabel;
  final bool needsRestock;
  final String? imagePath;
  final String? imageUrl;
  final DateTime? createdAt;
  final DateTime? updatedAt;
}
